File: tests/core/tools/codac2_tests_trunc.py
#!/usr/bin/env python

#  Codac tests
# ----------------------------------------------------------------------------
#  \date       2024
#  \author     Simon Rohou
#  \copyright  Copyright 2024 Codac Team
#  \license    GNU Lesser General Public License (LGPL)

import unittest
from codac import *
import sys

class TestTrunc(unittest.TestCase):

  def test_trunc(self):

    l = next_float(-oo); u = prev_float(oo)
    self.assertTrue(trunc(0) == 0)
    self.assertTrue(trunc(-oo) == l)
    self.assertTrue(trunc(oo) == u)
    self.assertTrue(untrunc(l) == -oo)
    self.assertTrue(untrunc(u) == oo)
    self.assertTrue(untrunc(0) == 0)
    
    # trunc matrix types

    self.assertTrue(trunc(IntervalMatrix([
        [[-1,1],[-2,2]],
        [[-3,3],[-4,4]]
      ])) == IntervalMatrix([
        [[-1,1],[-2,2]],
        [[-3,3],[-4,4]]
      ]))
    
    self.assertTrue(trunc(IntervalMatrix([
        [[-1,oo],[-2,2]],
        [[-3,3],[-4,4]]
      ])) == IntervalMatrix([
        [[-1,u],[-2,2]],
        [[-3,3],[-4,4]]
      ]))
    
    self.assertTrue(trunc(IntervalMatrix([
        [[-oo,oo],[-2,2]],
        [[-3,3],[-4,4]]
      ])) == IntervalMatrix([
        [[l,u],[-2,2]],
        [[-3,3],[-4,4]]
      ]))
    
    self.assertTrue(trunc(IntervalMatrix([
        [[-oo,oo],[-oo,oo]],
        [[-oo,oo],[-oo,oo]]
      ])) == IntervalMatrix([
        [[l,u],[l,u]],
        [[l,u],[l,u]]
      ]))
    
    # untrunc matrix types

    self.assertTrue(untrunc(IntervalMatrix([
        [[-1,1],[-2,2]],
        [[-3,3],[-4,4]]
      ])) == IntervalMatrix([
        [[-1,1],[-2,2]],
        [[-3,3],[-4,4]]
      ]))
    
    self.assertTrue(untrunc(IntervalMatrix([
        [[-1,u],[-2,2]],
        [[-3,3],[-4,4]]
      ])) == IntervalMatrix([
        [[-1,oo],[-2,2]],
        [[-3,3],[-4,4]]
      ]))
    
    self.assertTrue(untrunc(IntervalMatrix([
        [[l,u],[-2,2]],
        [[-3,3],[-4,4]]
      ])) == IntervalMatrix([
        [[-oo,oo],[-2,2]],
        [[-3,3],[-4,4]]
      ]))
    
    self.assertTrue(untrunc(IntervalMatrix([
        [[l,u],[l,u]],
        [[l,u],[l,u]]
      ])) == IntervalMatrix([
        [[-oo,oo],[-oo,oo]],
        [[-oo,oo],[-oo,oo]]
      ]))

if __name__ ==  '__main__':
  unittest.main()

File: tests/core/tools/codac2_tests_trunc.cpp
/** 
 *  Codac tests
 * ----------------------------------------------------------------------------
 *  \date       2024
 *  \author     Simon Rohou
 *  \copyright  Copyright 2024 Codac Team
 *  \license    GNU Lesser General Public License (LGPL)
 */

#include <catch2/catch_test_macros.hpp>
#include <codac2_IntervalMatrix.h>
#include <codac2_trunc.h>

using namespace std;
using namespace codac2;

TEST_CASE("trunc")
{
  double l = next_float(-oo), u = prev_float(oo);
  CHECK(codac2::trunc(0) == 0);
  CHECK(codac2::trunc(-oo) == l);
  CHECK(codac2::trunc(oo) == u);
  CHECK(untrunc(l) == -oo);
  CHECK(untrunc(u) == oo);
  CHECK(untrunc(0) == 0);
  
  // trunc matrix types

  CHECK(trunc(IntervalMatrix({
      {{-1,1},{-2,2}},
      {{-3,3},{-4,4}}
    })) == IntervalMatrix({
      {{-1,1},{-2,2}},
      {{-3,3},{-4,4}}
    }));
  
  CHECK(trunc(IntervalMatrix({
      {{-1,oo},{-2,2}},
      {{-3,3},{-4,4}}
    })) == IntervalMatrix({
      {{-1,u},{-2,2}},
      {{-3,3},{-4,4}}
    }));
  
  CHECK(trunc(IntervalMatrix({
      {{-oo,oo},{-2,2}},
      {{-3,3},{-4,4}}
    })) == IntervalMatrix({
      {{l,u},{-2,2}},
      {{-3,3},{-4,4}}
    }));
  
  CHECK(trunc(IntervalMatrix({
      {{-oo,oo},{-oo,oo}},
      {{-oo,oo},{-oo,oo}}
    })) == IntervalMatrix({
      {{l,u},{l,u}},
      {{l,u},{l,u}}
    }));
  
  // untrunc matrix types

  CHECK(untrunc(IntervalMatrix({
      {{-1,1},{-2,2}},
      {{-3,3},{-4,4}}
    })) == IntervalMatrix({
      {{-1,1},{-2,2}},
      {{-3,3},{-4,4}}
    }));
  
  CHECK(untrunc(IntervalMatrix({
      {{-1,u},{-2,2}},
      {{-3,3},{-4,4}}
    })) == IntervalMatrix({
      {{-1,oo},{-2,2}},
      {{-3,3},{-4,4}}
    }));
  
  CHECK(untrunc(IntervalMatrix({
      {{l,u},{-2,2}},
      {{-3,3},{-4,4}}
    })) == IntervalMatrix({
      {{-oo,oo},{-2,2}},
      {{-3,3},{-4,4}}
    }));
  
  CHECK(untrunc(IntervalMatrix({
      {{l,u},{l,u}},
      {{l,u},{l,u}}
    })) == IntervalMatrix({
      {{-oo,oo},{-oo,oo}},
      {{-oo,oo},{-oo,oo}}
    }));
}